setup-osx.rb

```diff
@@ -0,0 +1,183 @@
+class Installer
+  def self.full_install
+    ruby_bot = Installer.new
+    ruby_bot.ensure_bash_profile
+    [
+      'git', 'hub',
+      'npm', 'bower',
+      'psql',
+      'rbenv', 'ruby-build',
+      'bundler', 'foreman', 'ember-cli',
+      'cask',
+      'VirtualBox', 'boot2docker', 'docker-compose'
+    ].each do |program|
+      ruby_bot.install_program(program)
+    end
+  end
+
+  def ensure_bash_profile
+    bash_profile_location = '~/.bash_profile'
+    unless File.exist?(bash_profile_location)
+      File.write(File.expand_path(bash_profile_location), "")
+    end
+  end
+
+  def install_program(program, noisy=true)
+    if program_exists?(program)
+      puts "#{program} is already installed" if noisy
+      run_update_script(program, noisy)
+    else
+      puts "Installing #{program}" if noisy
+      check_prerequisites(program, noisy)
+      run_install_script(program, noisy)
+    end
+  end
+
+  protected
+
+  def program_exists?(program)
+    case program
+    when 'cask'
+      system("brew list brew-cask > /dev/null 2>&1")
+    when 'VirtualBox'
+      system("brew cask list virtualbox > /dev/null 2>&1")
+    when 'psql'
+      system("brew cask list postgres > /dev/null 2>&1")
+    when 'ember-cli', 'bower'
+      system("npm list -g --depth=0 #{program} > /dev/null 2>&1")
+    else
+      system("which #{ program} > /dev/null 2>&1")
+    end
+  end
+
+  def check_prerequisites(program, noisy=true)
+    find_prerequisites(program).each do |program|
+      install_program(program, false)
+    end
+  end
+
+  def run_install_script(program, noisy=true)
+    case program
+    when 'git'
+      system('brew update') # TODO Move this check
+      system('brew install git')
+    when 'brew'
+      system('ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"')
+    when 'bower'
+      system('npm install -g bower')
+    when 'npm'
+      system('curl https://raw.githubusercontent.com/creationix/nvm/v0.24.1/install.sh | bash') # TODO find a way to figure out latest version
+      system('nvm install stable')
+    when 'ruby-build'
+      system('brew install ruby-build')
+    when 'rbenv'
+      system('brew install rbenv')
+    when 'psql'
+      system('brew cask install postgres --appdir=/Applications')
+      system('brew install postgres')
+    when 'cask'
+      system('brew install caskroom/cask/brew-cask')
+    when 'hub'
+      system('brew install hub')
+    when 'boot2docker'
+      system('brew install boot2docker')
+    when 'docker-compose'
+      system('brew install docker-compose')
+    when 'VirtualBox'
+      system('brew cask install virtualbox')
+    when 'bundler'
+      system('gem install bundler')
+    when 'foreman'
+      system('gem install foreman')
+    when 'ember-cli'
+      system('npm install -g ember-cli')
+    else
+      # TODO?
+    end
```

Comment:
This is fine, but it might be more readable as a hash:
```rb
COMMANDS = {
  'ember-cli' => lambda { system('npm install -g ember-cli' },
  ...
}

command = COMMANDS[program]

if command
  command.call
else
  raise "Command #{program} does not exist"
end
```